src/openforms/formio/components/custom.py

```diff
@@ -639,6 +644,139 @@ def as_json_schema(component: AddressNLComponent) -> JSONObject:
         return base
 
 
+class PartnerListSerializer(serializers.ListSerializer):
+    def __init__(self, *args, **kwargs):
+        self.component = kwargs.pop("component", None)
+        super().__init__(*args, **kwargs)
+
+    def validate(self, attrs):
+        attrs = super().validate(attrs)
+
+        component_key = self.component["key"]
+        submission = self.context["submission"]
+        prefill_data = submission.get_prefilled_data()
+
+        fm_immutable_variables = FormVariable.objects.filter(
+            source=FormVariableSources.user_defined,
+            prefill_plugin=FM_PLUGIN_IDENTIFIER,
+        ).exclude(prefill_options={})
```

Comment:
```suggestion
        fm_immutable_variable = FormVariable.objects.filter(
            source=FormVariableSources.user_defined,
            prefill_plugin=FM_PLUGIN_IDENTIFIER,
            prefill_options__mutable_data_form_variable=component_key,
        ).first()
```

1. You can query the relevant variable directly, using the target component key
2. We expect only a single component to map to the variable we're validating (if not -> detect this in the digest checks)

This should replace the entire loop that follows :)